protoarray/src/main/java/tech.pegasys.artemis.protoarray/ProtoArrayForkChoice.java

```diff
@@ -0,0 +1,353 @@
+            justifiedEpoch,
+            finalizedEpoch,
+            new ArrayList<>(),
+            new HashMap<>());
+
+    protoArray.onBlock(
+        finalizedBlockSlot,
+        finalizedBlockRoot,
+        Optional.empty(),
+        finalizedBlockStateRoot,
+        justifiedEpoch,
+        finalizedEpoch);
+
+    return new ProtoArrayForkChoice(
+        protoArray, new ElasticList<>(VoteTracker::Default), new ArrayList<>());
+  }
+
+  void processAttestation(int validatorIndex, Bytes32 blockRoot, UnsignedLong targetEpoch) {
+    VoteTracker vote = votes.get(validatorIndex);
+
+    if (targetEpoch.compareTo(vote.getNextEpoch()) > 0 || vote.equals(VoteTracker.Default())) {
+      vote.setNextRoot(blockRoot);
+      vote.setNextEpoch(targetEpoch);
+    }
+  }
+
+  void processBlock(
+      UnsignedLong blockSlot,
+      Bytes32 blockRoot,
+      Bytes32 parentRoot,
+      Bytes32 stateRoot,
+      UnsignedLong justifiedEpoch,
+      UnsignedLong finalizedEpoch) {
+    protoArrayLock.writeLock().lock();
+    try {
+      protoArray.onBlock(
+          blockSlot, blockRoot, Optional.of(parentRoot), stateRoot, justifiedEpoch, finalizedEpoch);
+    } finally {
+      protoArrayLock.writeLock().unlock();
+    }
+  }
+
+  Bytes32 findHead(
+      UnsignedLong justifiedEpoch,
+      Bytes32 justifiedRoot,
+      UnsignedLong finalizedEpoch,
+      List<UnsignedLong> justifiedStateBalances) {
+    protoArrayLock.writeLock().lock();
+    votesLock.writeLock().lock();
+    balancesLock.writeLock().lock();
+    try {
+      List<UnsignedLong> oldBalances = balances;
+      List<UnsignedLong> newBalances = justifiedStateBalances;
+
+      List<Long> deltas = computeDeltas(protoArray.getIndices(), votes, oldBalances, newBalances);
+
+      protoArray.applyScoreChanges(deltas, justifiedEpoch, finalizedEpoch);
+      balances = new ArrayList<>(newBalances);
+
+      return protoArray.findHead(justifiedRoot);
+    } finally {
+      protoArrayLock.writeLock().unlock();
+      votesLock.writeLock().unlock();
+      balancesLock.writeLock().unlock();
+    }
+  }
+
+  public void setPruneThreshold(int pruneThreshold) {
+    protoArrayLock.writeLock().lock();
+    try {
+      protoArray.setPruneThreshold(pruneThreshold);
+    } finally {
+      protoArrayLock.writeLock().unlock();
+    }
+  }
+
+  public int size() {
+    protoArrayLock.readLock().lock();
+    try {
+      return protoArray.getNodes().size();
+    } finally {
+      protoArrayLock.readLock().unlock();
+    }
+  }
+
+  public boolean containsBlock(Bytes32 blockRoot) {
+    protoArrayLock.readLock().lock();
+    try {
+      return protoArray.getIndices().containsKey(blockRoot);
+    } finally {
+      protoArrayLock.readLock().unlock();
+    }
+  }
+
+  public Optional<UnsignedLong> blockSlot(Bytes32 blockRoot) {
+    return blockSlotAndStateRoot(blockRoot).map(BlockSlotAndStateRoot::getBlockSlot);
+  }
+
+  public Optional<BlockSlotAndStateRoot> blockSlotAndStateRoot(Bytes32 blockRoot) {
```

Comment:
We turned out to not use any of these functions.  Is there something we should be doing with them?

Reply:
Made a JIRA ticket as discussed offline: https://pegasys1.atlassian.net/secure/RapidBoard.jspa?rapidView=67&modal=detail&selectedIssue=BC-344